### PR TITLE
hostapd: add ppsk option (support for private psk)

### PR DESCRIPTION
**Pull request by mgiganto**
This PR allows a user to enable a private psk, where each station
may have it's own psk or use a common psk if it is not defined.
The private psk is defined using the sta's mac and a radius server
is required.

ppsk option should be enabled in the wireless configuration along with
radius server details. When using PPSK, the key is ignored, it will be
retrieved from radius server. SAE is not yet supported (private sae) in
hostapd.

Wireless example configuration:
	option encryption 'psk2+ccmp'
	option ppsk '1'
	option auth_server '127.0.0.1'
	option auth_secret 'radiusServerPassword'

If you want to use dynamic VLAN on PPSK also include:
	option dynamic_vlan '2'
	option vlan_tagged_interface 'eth0'
	option vlan_bridge 'br-vlan'
	option vlan_naming '0'

It works enabling mac address verification on radius server and
requiring the tunnel-password (the private psk) from radius server.

In the radius server we need to configure the users. In case of
freeradius: /etc/freeradius3/mods-config/files/authorize
The user and Cleartext-Password should be the mac lower case using the
format "aabbccddeeff"

<sta mac> Cleartext-Password := "<sta mac>"
	Tunnel-Password = <Private Password>

Example of a user configured in radius and using dynamic VLAN5:

8cb84a000000 Cleartext-Password := "8cb84a000000"
	Tunnel-Type = VLAN,
	Tunnel-Medium-Type = IEEE-802,
	Tunnel-Private-Group-ID = 5,
	Tunnel-Password = MyPrivPw

If we want to have a default or shared psk, used when the mac is not
found in the list, we need to add the following at the end of the radius
authorize file:

DEFAULT Auth-Type := Accept
	Tunnel-Password = SharedPw

And if using VLANs, for example VLAN6 for default users:
DEFAULT Auth-Type := Accept
	Tunnel-Type = VLAN,
	Tunnel-Medium-Type = IEEE-802,
	Tunnel-Private-Group-ID = 6,
	Tunnel-Password = SharedPw

Signed-off-by: Manuel Giganto <mgigantoregistros@gmail.com>

### CODE REVIEW

File: package/network/services/hostapd/files/hostapd.sh [hunk 1]
```diff
@@ -484,7 +485,15 @@ hostapd_set_bss_options() {
 		;;
 		psk|sae|psk-sae)
 			json_get_vars key wpa_psk_file
-			if [ ${#key} -eq 64 ]; then
+			if [ "$auth_type" == "psk" ] && [ "$ppsk" -ne 0 ] ; then
```
Comment (adschm) [inline package/network/services/hostapd/files/hostapd.sh:488]: Please use single "=" instead of bash "=="

Comment (mgiganto) [reply package/network/services/hostapd/files/hostapd.sh]: Done

**Comment by ynezz** [comment]
Please rebase

**Comment by CodeFetch** [comment]
@mgiganto Great work! I've been waiting for this. Do you have a clue why private SAE is not supported, yet?

**Comment by ynezz** [comment]
Ping.

**Comment by adschm** [comment]
@mgiganto Ping.

File: package/network/services/hostapd/files/hostapd.sh [hunk 2]
```diff
@@ -484,7 +485,15 @@ hostapd_set_bss_options() {
...
 			json_get_vars key wpa_psk_file
-			if [ ${#key} -eq 64 ]; then
+			if [ "$auth_type" = "psk" ] && [ "$ppsk" -ne 0 ] ; then
+				json_get_vars auth_server auth_secret auth_port
+				set_default auth_port 1812
+				append bss_conf "auth_server_addr=$auth_server" "$N"
+				append bss_conf "auth_server_port=$auth_port" "$N"
+				append bss_conf "auth_server_shared_secret=$auth_secret" "$N"
+				append bss_conf "macaddr_acl=2" "$N"
+				append bss_conf "wpa_psk_radius=2" "$N"
```
Comment (bchavet) [inline package/network/services/hostapd/files/hostapd.sh:495]: As eluded to in https://github.com/openwrt/openwrt/pull/3509#issuecomment-749548980, this PR is a bit behind the current stable version. After manually applying this diff to hostapd.sh from v19.07.6, I also had to add the following line here in order to make dynamic VLANs work

```
vlan_possible=1
```

Afterwards, this works beautifully!

**Comment by adschm** [comment]
This is abandoned.

**Comment by mgiganto** [comment]
Sorry, I didn't see the email... pong

**Comment by mgiganto** [comment]
The PR is now rebased @ynezz.

File: package/network/services/hostapd/files/hostapd.sh (hunk 2 above)
Comment (mgiganto) [reply package/network/services/hostapd/files/hostapd.sh:588]: Yes, if you are on 19.07 or older you will have to do vlan_possible=1, master has it already at the end of the case statement and it is not required. Thank you @bchavet for pointing it out for other people that may be interested in using it outside of master :D

**Comment by mgiganto** [comment]
> 
> 
> @mgiganto Great work! I've been waiting for this. Do you have a clue why private SAE is not supported, yet?

I'll try/research it again, when I did this PR, hostapd didn't support ppsk over SAE, I hope hostapd support it soon, so we can use PSAE :)

**Comment by CodeFetch** [comment]
@mgiganto I'm planning to use this in community mesh networks. The nodes can't trust other nodes. Therefore they should use different PSKs. If this works with SAE that would be great. I would then implement a minimal Radius server that handles the keys. Which radius server do you use for your tests?

**Comment by mgiganto** [comment]
@CodeFetch SAE is not yet supported because hostapd doesn't supported it yet unfortunately, looks like they are doing a lot of improvements in SAE and it may be supported soon.

For the radius server, I used freeradius, included in openwrt so you can run it directly on each node (remember that each node has to validate credentials of the sta connecting to them, so it will require to connect to radius to validate credentials.

You should be able to use any radius server, including the one provided by windows or the hostapd internal radius system with custom config, but I found it more confortable and easy to maintain if I use the freeradius running locally.
If your nodes are all connected by cable or very stable network, you can relay in one single radius server running in an always-available place for the APs, otherwise run it locally on each.

**Comment by andrey-yantsen** [comment]
@ynezz, sorry for asking a stupid question, but is there anything I can do to help get this merged? It's been a while since the PR was created, and I'd really love to see the changes in master.

**Comment by andrey-yantsen** [comment]
@adschm maybe you can advise on how to get this merged?

**Comment by andrey-yantsen** [comment]
@mgiganto, can you please resolve the conflict? Before I start trying to invite other maintainers who might help us :)

**Comment by mgiganto** [comment]
Merge conflict solved again. Thank you @andrey-yantsen  and Neustradamos!

P.S. I hope soon I will have another feature related with dynamic security/isolation/IoT that may require to patch hostapd. 
I wonder what it would be best way to merge it "fast", send it as a patch or as a PR or maybe both.

**Comment by andrey-yantsen** [comment]
@nbd168 sorry for involving you in this, but I saw a few recent commits from you to hostapd. Would you mind helping to get this merged or give a piece of advice on what needs to be done to get it merged?

**Comment by andrey-yantsen** [comment]
Hey @blocktrron! Sorry for dragging you into this too, but it seems like everybody is busy :( I saw a few merged PRs from you into hostapd, and maybe you can give some advice here on how we can get this PR merged?

**Comment by falaca** [comment]
Thanks for writing this patch @mgiganto - I'm looking forward to using this so I hope it gets merged soon.

**Comment by tricksol** [comment]
Can we please get some reviewers to look at this?

**Comment by mgiganto** [comment]
It is on core developers team.
You can ask them on irc :)

On Tuesday, 30 November 2021, Trick Solutions ***@***.***>
wrote:
> Can we please get some reviewers to look at this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
> Triage notifications on the go with GitHub Mobile for iOS or Android. <
https://ci4.googleusercontent.com/proxy/1HGDm7coCI0nIKlBzYJmBeCykIcW9-P29gPWh091wIiWCjIQhl--Ha9iOOSEaMJPl5o5_IbL34V94CzitXPlyi2Vg03LqVKfVLYAYWrRXS7dL5SMvNpF3IEEiVb_FyPa2ZQvFPXnDDRlB5NQ5cgy0gL2Ng6V5x8wQ07ErVOQ8qrieL7bNcILdYTnlkQ-kEg4FPunu9rlmagWPmNA23QaE8Ffpg5Zxo77-CkSlTU6tw=s0-d-e1-ft#https://github.com/notifications/beacon/AFB3LFYYX2JCIOO4H3WAA7DUOT333A5CNFSM4SMURCM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOHKKGN5Y.gif>

**Comment by moham96** [comment]
When will this get merged?

**Comment by ynezz** [comment]
> @ynezz, sorry for asking a stupid question, but is there anything I can do to help get this merged? It's been a while since the PR was created, and I'd really love to see the changes in master.

Indeed, test it and give me your `Tested-by:` tag.

@mgiganto Can you please rebase?

**Comment by Ansuel** [comment]
@mgiganto news on this?

**Comment by Ansuel** [comment]
@mgiganto merged in master with d12eb103e875ea85c2f1e9a2e9a2922859ec9896

Also you should update the wiki with this new option and also luci to add support for this!

**Comment by mgiganto** [comment]
> @mgiganto merged in master with [d12eb10](https://github.com/openwrt/openwrt/commit/d12eb103e875ea85c2f1e9a2e9a2922859ec9896)
> 
> Also you should update the wiki with this new option and also luci to add support for this!

Excellent news!!!
Thank you very much.

I´ll update the luci PR and mention you when it is ready.

**Comment by garlett** [comment]
tested and is working on OpenWrt SNAPSHOT r22297-1781e7408a ;

but there are other options that do not require radius:
1) multiple "keyid=11 00:00:00:00:00:00 11223344" lines on default "/tmp/run/hostapd-phy1-ap1.psk" location
- change location with "option wpa_psk_file '/etc/hostapd_wpa.psk'" on /etc/config/wireless
- keyid is logged on AP-STA-CONNECTED event, I use to it on a cron script to auto-bind client mac with psk
- 00:00:00:00:00:00 accept any mac
- run "ubus call hostapd.phy1-ap1 reload" to reload;

2) multiple "sae_password=11223344|mac=00:00:00:00:00:00" lines on hostapd.conf ( did not worked for me );

**Comment by mgiganto** [comment]
SAE is not supported

**Comment by milindpatel63** [comment]
> This PR allows a user to enable a private psk, where each station may have it's own psk or use a common psk if it is not defined. The private psk is defined using the sta's mac and a radius server is required.
> 
> ppsk option should be enabled in the wireless configuration along with radius server details. When using PPSK, the key is ignored, it will be retrieved from radius server. SAE is not yet supported (private sae) in hostapd.
> 
> Wireless example configuration: option encryption 'psk2+ccmp' option ppsk '1' option auth_server '127.0.0.1' option auth_secret 'radiusServerPassword'
> 
> If you want to use dynamic VLAN on PPSK also include: option dynamic_vlan '2' option vlan_tagged_interface 'eth0' option vlan_bridge 'br-vlan' option vlan_naming '0'
> 
> It works enabling mac address verification on radius server and requiring the tunnel-password (the private psk) from radius server.
> 
> In the radius server we need to configure the users. In case of freeradius: /etc/freeradius3/mods-config/files/authorize The user and Cleartext-Password should be the mac lower case using the format "aabbccddeeff"
> 
> Cleartext-Password := "" Tunnel-Password =
> 
> Example of a user configured in radius and using dynamic VLAN5:
> 
> 8cb84a000000 Cleartext-Password := "8cb84a000000" Tunnel-Type = VLAN, Tunnel-Medium-Type = IEEE-802, Tunnel-Private-Group-ID = 5, Tunnel-Password = MyPrivPw
> 
> If we want to have a default or shared psk, used when the mac is not found in the list, we need to add the following at the end of the radius authorize file:
> 
> DEFAULT Auth-Type := Accept Tunnel-Password = SharedPw
> 
> And if using VLANs, for example VLAN6 for default users: DEFAULT Auth-Type := Accept Tunnel-Type = VLAN, Tunnel-Medium-Type = IEEE-802, Tunnel-Private-Group-ID = 6, Tunnel-Password = SharedPw
> 
> Signed-off-by: Manuel Giganto [mgigantoregistros@gmail.com](mailto:mgigantoregistros@gmail.com)

Can we have multiple Defaults in freeradius, which return different vlan id based on Tunnel-Password?
I have tried Fall Through = Yes, but still freeradius only returns the first one, and if first one fails with wrong Tunnel-Password, it's not going to next one.

**Pull request closed without merging**